src/test/completionParticipant.test.ts

```diff
@@ -0,0 +1,123 @@
+
+  let assertCompletion = function (completions: CompletionList, expected: ItemDescription, document: TextDocument, offset: number) {
+		let matches = completions.items.filter(completion => {
+			return completion.label === expected.label;
+		});
+		if (expected.notAvailable) {
+			assert.equal(matches.length, 0, expected.label + " should not existing is results");
+			return;
+		}
+
+		assert.equal(matches.length, 1, expected.label + " should only existing once: Actual: " + completions.items.map(c => c.label).join(', '));
+		let match = matches[0];
+		if (expected.documentation) {
+			assert.equal(match.documentation, expected.documentation);
+		}
+		if (expected.kind) {
+			assert.equal(match.kind, expected.kind);
+		}
+		if (expected.resultText && match.textEdit) {
+			assert.equal(applyEdits(document, [match.textEdit]), expected.resultText);
+		}
+	};
+
+	let testCompletionWithParticipantsFor= function (value: string, participants: htmlLanguageService.ICompletionParticipant[]): PromiseLike<void> {
+		let offset = value.indexOf('|');
+		value = value.substr(0, offset) + value.substr(offset + 1);
+
+		let ls = htmlLanguageService.getLanguageService();
+
+		let document = TextDocument.create('test://test/test.html', 'html', 0, value);
+		let position = document.positionAt(offset);
+    let htmlDoc = ls.parseHTMLDocument(document);
+    ls.setCompletionParticipants(participants);
+    let list = ls.doComplete(document, position, htmlDoc);
+
+		return Promise.resolve();
+  };
+  
+  test('onHtmlAttributeValue', () => {
+    let attributeValueParticipationCount = 0;
+    let tag, attribute, value;
+    let range: Range = Range.create(Position.create(0, 0), Position.create(0, 0));
+
+    const participant: htmlLanguageService.ICompletionParticipant = {
+      onHtmlAttributeValue: (context: { tag: string, attribute: string; value: string; range: Range }) => {
+        tag = context.tag;
+        attribute = context.attribute;
+        value = context.value;
+        range = context.range;
+        attributeValueParticipationCount++;
+      }
+    };
+    testCompletionWithParticipantsFor('<|', [participant]);
+    testCompletionWithParticipantsFor('<div |>', [participant]);
+    testCompletionWithParticipantsFor('<div class|>', [participant]);
+    testCompletionWithParticipantsFor('<div>|', [participant]);
+    testCompletionWithParticipantsFor('<div>|</div>', [participant]);
+    assert.equal(attributeValueParticipationCount, 0);
+
+    testCompletionWithParticipantsFor('<div class="|"></div>', [participant]);
+    assert.equal(attributeValueParticipationCount, 1);
+    assert.equal(tag, 'div');
+    assert.equal(attribute, 'class');
+    assert.equal(value, '');
+    assert.equal(range.start.line, 0);
+    assert.equal(range.start.character, 12);
+    assert.equal(range.end.line, 0);
+    assert.equal(range.end.character, 12);
+
+    testCompletionWithParticipantsFor('<div class="foo|"></div>', [participant]);
```

Comment:
What if the cursor is before or after the `"`? What range is returned then?

It seems to me the range should always include the quotes.

Reply:
I think this is in-line with the `onHtmlAttributeValue` where they only receive the unquoted value, as I feel the participant shouldn't need to worry about the quotes outside.

If you think it's better to have the quotes, I'll change both the test and the `value` in `onHtmlAttributeValue`, and do the extraction of the surrounding quotes in HTML server's `pathCompletion`.

Reply:
I think it's better to always get the full range, including the quotes. That way the completion particpant can also decide to remove the quotes. The 'value' should be without quotes

Reply:
I did it this way. However the previous behavior is that for `<div class='foo'>`, with `scanner.getTokenText()` you seem to get normalized value with double quotes (`"foo"`). But that's for another issue I guess.